gobblin-runtime/src/main/java/gobblin/runtime/MultiConverter.java

```diff
@@ -80,82 +78,111 @@ public Object convertSchema(Object inputSchema, WorkUnitState workUnit)
-              converter.convertRecord(convertedSchemaMap.get(converter), record, workUnitState).iterator();
+      // Set the new value of currentRecord
+      setNextRecord();
+    }
+
+    /**
+     * Helper method to set the newest value of currentRecord, after this method is called then it is safe to return the
+     * value of currentRecord. The method works by propagating the value currentRecord up the list until all converters
+     * have been reset based on the value of currentRecord. If the propagation ever finds an EmptyIterable, then it
+     * moves to the next available element in the list.
+     * @throws DataConversionException
+     */
+    public void setNextRecord() throws DataConversionException {
+      while (this.currentRecord != null && this.converterIteratorStack.size() != converters.size()) {
+        Converter converter = converters.get(converterIteratorStack.size());
+        Iterator<Object> iterator =
+            converter.convertRecord(convertedSchemaMap.get(converter), this.currentRecord, this.workUnitState)
+                .iterator();
+
+        if (iterator.hasNext()) {
           this.converterIteratorStack.push(iterator);
-          // Do not continue when encountering an empty iterator
-          if (!iterator.hasNext()) {
-            break;
-          }
+          this.currentRecord = iterator.next();
+
+        } else {
+          this.currentRecord = getNextElementFromStack();
         }
       }
+    }
 
-      this.outputIterator = this.converterIteratorStack.peek();
+    /**
+     * Helper method to get the next element on the list. This does a search down the stack checking each iterator to
+     * see if any iterator has an element. Once it finds that element, it returns it.
+     * @returns the next element from the stack
+     */
+    public Object getNextElementFromStack() {
+      Iterator<Object> itr;
+
+      while (!this.converterIteratorStack.isEmpty()) {
+        itr = this.converterIteratorStack.peek();
+        if (itr.hasNext()) {
+          return itr.next();
+        } else {
+          this.converterIteratorStack.pop();
+        }
+      }
+      return null;
     }
 
     @Override
     public boolean hasNext() {
-      if (this.outputIterator.hasNext()) {
-        return true;
-      }
+      return this.currentRecord != null;
+    }
 
-      // Pop out all empty converter iterators from the stack
-      while (!this.converterIteratorStack.isEmpty() && !this.converterIteratorStack.peek().hasNext()) {
-        this.converterIteratorStack.pop();
-      }
+    @Override
+    public Object next() {
+
+      // The returnRecord will be the record that is returned by this method
+      Object returnRecord = this.currentRecord;
 
+      // Move the currentRecord point to the next record, or null if there are no more records left
       if (this.converterIteratorStack.isEmpty()) {
-        return false;
+        this.currentRecord = null;
+        return returnRecord;
       }
 
```

Comment:
Please add an assert that this.converterIteratorStack.size () == this.converters.size()